### PR TITLE
#37 Enable notifications

### PR DESCRIPTION
**Pull request by alexjgriffith**
`mastodon-notifications--get` will now create a `*mastodon-notifications*` buffer.

Four types of notifications have rendering functions
1. mention
2. follow
3. favourite
4. reblog (boost)

Mention, favourite, and boost are very similar to standard toots. Follows will have a tag line (which could be customizable) and will insert the image of the new follower before the end of the notification. 
## Mention
```
$content
 | $display_name (@$user_name) Mentioned $your_display_name(@$your_user_name)
  ------------
```

## Follow
```
Congratulations, you have a new follower!
;; dependent on the inclusion of  mastodon-media.el 
;; $user_image
 | $display_name (@$user_name) Following $your_display_name(@$your_user_name)
  ------------
```
## Favourite
```
$content
 | $display_name (@$user_name)  Favorited $your_display_name(@$your_user_name)
  ------------
```

## Booseted
```
$content
 | $display_name (@$user_name)  Boosted $your_display_name(@$your_user_name)
  ------------
```

The functionality will be in the `mastodon-notifications` module and will depend explicitly on `mastodon-media`, `mastodon-http` and `mastodon-tl`

A lot of the utility overlaps with those `mastdodon-tl`. For the time being I am using the functionality in from the `mastodon-tl` package for navigation, however a more generalized version of  updating functions like `mastodon-tl--more` will be required going forward. 

## Example from Emacs
notifications buffer

### CODE REVIEW

**Comment by alexjgriffith** [comment]
More and Update are currently broken due to the way `mentions` are handled. They insert the id of the toot being mentioned rather than the id of the notification.

File: lisp/mastodon-notifications.el [hunk 1]
```diff
@@ -0,0 +1,168 @@
+;;; mastodon-notifications.el --- Notification functions for mastodon.el
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: No `lexical-binding: t` ?

File: lisp/mastodon-notifications.el [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+;; Copyright (C) 2017 Johnson Denen
+;; Author: Johnson Denen <johnson.denen@gmail.com>
+;; Version: 0.5.4
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: "0.5.4"?  oh no, it isn't

File: lisp/mastodon-notifications.el [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+;;       - replace `You' with `your_display_name'(@`your_user_name')
+;;       - generalize mastodon-tl--more or rebind it in notifications
+;;         example
+;;       - fix thread access
+;;       - Add keybinding of mastodon-notifications--get to `N'
+
+;; ;; One approach for generalizing the get-more functionality
+;; (defun mastodon-tl--get-more (more-function id-fun goto-fun
+;;                               processing-function)
+;;   "can work for both oldert toots and newest toots"
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: s/oldert/older/

File: lisp/mastodon-notifications.el [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+;;   (interactive)
+;;   (let* ((tl (mastodon-tl--timeline-name))
+;;       (id (funcall id-fun))
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: indentation is off

File: lisp/mastodon-notifications.el [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+;;       (json (funcall more-function tl id)))
+;;     (when json
+;;       (with-current-buffer (current-buffer)
+;;      (let ((inhibit-read-only t))
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: indentation is totally off...

File: lisp/mastodon-notifications.el [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+;;                        point-min mastodon-tl--timeline)
+
+
+;;; Code:
+
+;;(require 'mastodon-media)
+(require 'mastodon-http)
+(require 'mastodon-tl)
+
+(defgroup mastodon-notifications nil
```
Comment (hdurer) [inline lisp/mastodon-notifications.el:44]: Do we need this?

File: lisp/mastodon-notifications.el [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+             (mastodon-notifications--byline-concat toot message)
+             (propertize "\n  ------------" 'face 'default))
+     'toot-id id
+     'toot-json toot)))
+
+(defun mastodon-notifications--mention (note)
+  "Format for a `mention' NOTE."
+  (let ((toot (mastodon-tl--field 'status note)))
+    (insert
+     (concat (mastodon-tl--content toot)
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: `concat` not needed (we make that mistake in various places).

`insert` takes any amount of args and inserts them in order, AFAIK.

File: lisp/mastodon-notifications.el [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+;;(require 'mastodon-media)
+(require 'mastodon-http)
+(require 'mastodon-tl)
+
+(defgroup mastodon-notifications nil
+  "Mastodon Notifications."
+  :prefix "mastodon-notifications-"
+  :group 'mastodon)
+
+(defvar mastodon-notifications--types
```
Comment (hdurer) [inline lisp/mastodon-notifications.el:49]: Why not have a plist of type to handling-function?  Then you don't repeat yourself below in the dispatching function.

File: lisp/mastodon-notifications.el [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  "Filters NOTE for those listed in `mastodon-notifications--types'.
+
+  It then passes the results to the `caller'"
+  (let ((type (mastodon-tl--field 'type note)))
+    (when (member type mastodon-notifications--types)
+      (mastodon-notifications--caller type note))))
+
+(defun mastodon-notifications--notifications (json)
+  "Format JSON in Emacs buffer."
+  (mapcar #'mastodon-notifications--note json)
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: Use `mapc` if you map only for the side effects.

File: lisp/mastodon-notifications.el [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  (replace-regexp "\n\n\n | " "\n | " nil (point-min) (point-max)))
```
Comment (hdurer) [inline lisp/mastodon-notifications.el]: IIRC the linter doesn't like using this in elisp code.

Comment (hdurer) [inline lisp/mastodon-notifications.el]: I've never got to the bottom of this - why do we first insert these newlines only to later remove them again?

File: lisp/mastodon-notifications.el (hunk 8 above)
Comment (alexjgriffith) [reply lisp/mastodon-notifications.el:49]: Fixed, turned into plist

**Comment by alexjgriffith** [comment]
## Changes
- Removed leading sample code
- Removed redundant `concat`s within `insert`s
- made `mastodon-notifications--types` into a plist
- pulled `fixed-numberp` in order to make sure updating and loading more notifications works
- pulled develop

**Comment by alexjgriffith** [comment]
This will be revisited later

**Pull request closed without merging**